### PR TITLE
Fixed lbfgs for ray-cluster

### PR DESCRIPTION
**Pull request by Wapaul1**

### CODE REVIEW

File: examples/lbfgs/driver.py [hunk 1]
```diff
@@ -27,40 +27,37 @@
...
-      for var in tf.trainable_variables():
-        assignment_placeholders.append(tf.placeholder(var.value().dtype, var.get_shape().as_list()))
-        assignment_nodes.append(var.assign(assignment_placeholders[-1]))
-
-      def get_weights():
-        return [v.eval(session=sess) for v in tf.trainable_variables()]
-
-      def set_weights(new_weights):
-        sess.run(assignment_nodes, feed_dict={p: w for p, w in zip(assignment_placeholders, new_weights)})
+  class LinearModel(object):
```
Comment (pcmoritz) [inline examples/lbfgs/driver.py]: Can you move this class definition before the __main__ section of the file (after the imports)? Same for net_initialization and net_reinitialization, loss, grad, full_loss and full_grad!

File: examples/lbfgs/driver.py [hunk 2]
```diff
@@ -27,40 +27,37 @@
...
+    def __init__(self, shape):
+      image_dimension = shape[0]
```
Comment (pcmoritz) [inline examples/lbfgs/driver.py]: why don't you get rid of image_dimension and label_dimension? In general LinearModel can be an arbitrary linear model (not related to images) and these could be arbitrary shapes, and just substituting shape[0] and shape[1] is good!

File: examples/lbfgs/driver.py [hunk 3]
```diff
@@ -73,27 +70,29 @@ def set_weights(new_weights):
...
-  # Create a reusable variable for the network.
+  # Register the network with Ray and create a reusable variable for it.
+  ray.register_class(LinearModel)
   ray.reusables.net_vars = ray.Reusable(net_initialization, net_reinitialization)
 
   # Load the weights into the network.
   def load_weights(theta):
-    sess, _, _, _, _, get_weights, set_weights = ray.reusables.net_vars
-    set_weights([theta[:w_size].reshape(w_shape), theta[w_size:].reshape(b_shape)])
+    net = ray.reusables.net_vars
```
Comment (pcmoritz) [inline examples/lbfgs/driver.py]: Please get rid of load_weights, that is the whole point of having TensorFlowVariables

File: examples/lbfgs/driver.py [hunk 4]
```diff
@@ -73,27 +70,29 @@ def set_weights(new_weights):
   def net_reinitialization(net_vars):
     return net_vars
 
-  # Create a reusable variable for the network.
+  # Register the network with Ray and create a reusable variable for it.
+  ray.register_class(LinearModel)
```
Comment (pcmoritz) [inline examples/lbfgs/driver.py]: register_class shouldn't be neccessary, we shouldn't have to serialize the network (and we shouldn't, if we have to there is a problem somewhere)

File: examples/lbfgs/driver.py [hunk 5]
```diff
@@ -73,27 +70,29 @@ def set_weights(new_weights):
...
   ray.reusables.net_vars = ray.Reusable(net_initialization, net_reinitialization)
 
   # Load the weights into the network.
   def load_weights(theta):
```
Comment (pcmoritz) [inline examples/lbfgs/driver.py]: can you define a generic function in tfutils.py which takes a dict and flattens it into a numpy array? It might be necessary to go from dicts to ordered dicts for that (and do ray.register_class(ordereddict) or similar)
